File: services/api/telemetry.mjs
import { NodeSDK } from '@opentelemetry/sdk-node';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { ConsoleSpanExporter } from '@opentelemetry/sdk-trace-base';
import { Resource } from '@opentelemetry/resources';

let sdk;

function createExporter() {
  const endpoint = process.env.OTEL_EXPORTER_OTLP_ENDPOINT;
  if (endpoint) {
    const url = `${endpoint.replace(/\/$/, '')}/v1/traces`;
    return new OTLPTraceExporter({ url });
  }
  return new ConsoleSpanExporter();
}

export function startTelemetry() {
  if (sdk) {
    return sdk;
  }

  const resource = new Resource({
    'service.name': process.env.OTEL_SERVICE_NAME ?? 'blackroad-api-bridge',
    'service.version': process.env.BUILD_SHA ?? 'dev',
  });

  sdk = new NodeSDK({
    resource,
    traceExporter: createExporter(),
    instrumentations: [getNodeAutoInstrumentations()],
  });

  sdk.start().catch((err) => {
    console.warn('[telemetry] failed to start OpenTelemetry SDK', err);
  });

  const shutdown = () => {
    sdk
      .shutdown()
      .catch((err) => console.warn('[telemetry] shutdown failed', err))
      .finally(() => process.exit(0));
  };

  process.once('SIGTERM', shutdown);
  process.once('SIGINT', shutdown);

  return sdk;
}

startTelemetry();

File: services/materials_service/telemetry.py
"""Telemetry bootstrap for the materials service."""

from __future__ import annotations

import os
from typing import Optional

from fastapi import FastAPI
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter

_configured = False


def _build_exporter() -> BatchSpanProcessor:
    endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
    if endpoint:
        url = f"{endpoint.rstrip('/')}/v1/traces"
        exporter = OTLPSpanExporter(endpoint=url)
    else:
        exporter = ConsoleSpanExporter()
    return BatchSpanProcessor(exporter)


def configure_telemetry(app: FastAPI) -> Optional[TracerProvider]:
    """Configure OpenTelemetry for the given FastAPI app."""
    global _configured
    if _configured:
        return trace.get_tracer_provider()

    resource = Resource.create(
        {
            "service.name": os.getenv("OTEL_SERVICE_NAME", "lucidia-materials"),
            "service.version": os.getenv("BUILD_SHA", "dev"),
        }
    )
    provider = TracerProvider(resource=resource)
    provider.add_span_processor(_build_exporter())
    trace.set_tracer_provider(provider)
    FastAPIInstrumentor.instrument_app(app)
    _configured = True
    return provider
